### PR TITLE
[execution] Update to deps-only mode 

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

Disallow new modules from being added to packages in deps-only mode. 

## Test plan 

Added new tests to check for both previous and new behavior.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [X] Protocol: Updated restrictions on deps-only packages to not allow adding new modules to the package.
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs [hunk 1]
```diff
@@ -457,6 +457,122 @@ async fn test_upgrade_package_compatible_in_dep_only_mode() {
...
+        .await;
+
+    assert!(effects.status().is_ok(), "{:#?}", effects.status());
+}
+
+#[tokio::test]
+async fn test_upgrade_package_add_new_module_in_dep_only_mode() {
+    let mut runner = UpgradeStateRunner::new("move_upgrade/base").await;
+
+    let (digest, modules) = build_upgrade_test_modules("dep_only_upgrade_new_module");
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs:487]: would it be possible to slice out the new module here, retry the publish, and see that it succeeds?

File: sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs [hunk 2]
```diff
@@ -660,6 +660,25 @@ mod checked {
             invariant_violation!("Tried to normalize modules in existing package but failed")
         };
 
+        let existing_modules_len = current_normalized.len();
+        let upgrading_modules_len = upgrading_modules.len();
+        if context
+            .protocol_config
+            .disallow_new_modules_in_deps_only_packages()
+            && policy as u8 == UpgradePolicy::DEP_ONLY
+            && existing_modules_len != upgrading_modules_len
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs:669]: Should we add some sort of invariant below about `new_normalized` being empty in the `DEP_ONLY` case if `disallow_new_modules_in_deps_only_packages`

Fine with it being `debug_only`

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs (hunk 1 above)
Comment (tzakian) [reply crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs:487]: We have that exact test pretty much here: https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs#L573

This is the same package but without `new_module.move`.

Comment (tnowacki) [reply crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs:487]: That is a slightly different from what I am asking. As it stands, it is not guaranteed that the error in this test from the fact that there is a new module. 
If you take the modules list and splice it out, you would guarantee that within this test.

File: sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs (hunk 2 above)
Comment (tzakian) [reply sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs:669]: I like it! Will add

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs (hunk 1 above)
Comment (tzakian) [reply crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs]: Updated so they all start from the valid `dep_only_upgrade`, and then explicitly add or remove files from that.

**Review by tnowacki — APPROVED** [review]

**Pull request merged**